storage/innobase/include/trx0purge.h

```diff
@@ -55,80 +55,80 @@ Run a purge batch.
-	/** Constructor */
-	TrxUndoRsegsIterator();
-	/** Sets the next rseg to purge in purge_sys.
-	Executed in the purge coordinator thread.
-	@retval false when nothing is to be purged
-	@retval true  when purge_sys.rseg->latch was locked */
-	inline bool set_next();
-
-private:
-	// Disable copying
-	TrxUndoRsegsIterator(const TrxUndoRsegsIterator&);
-	TrxUndoRsegsIterator& operator=(const TrxUndoRsegsIterator&);
-
-	/** The current element to process */
-	TrxUndoRsegs			m_rsegs;
-	/** Track the current element in m_rsegs */
-	TrxUndoRsegs::const_iterator	m_iter;
-};
-
 /** The control structure used in the purge operation */
 class purge_sys_t
 {
-  friend TrxUndoRsegsIterator;
+  /** Min-heap based priority queue over fixed size array */
+  class purge_queue
+  {
+    /** Array of indexes in trx_sys.rseg_array. */
+    alignas(CPU_LEVEL1_DCACHE_LINESIZE) byte m_array[TRX_SYS_N_RSEGS];
+    /** Pointer to the end of m_array. */
+    byte *m_end= m_array;
+
+  public:
+    struct trx_rseg_cmp
+    {
+      /** Compare two trx_rseg_t* based on trx_no.
+      @param lhs first index in trx_sys.rseg_array to compare
+      @param rhs second index in trx_sys.rseg_array to compare
+      @return whether lhs>rhs */
+      bool operator()(const byte lhs, const byte rhs)
+      {
+        ut_ad(lhs < TRX_SYS_N_RSEGS);
+        ut_ad(rhs < TRX_SYS_N_RSEGS);
+        /* We can compare without trx_rseg_t::latch, because rseg last
+        commit is always set before pushing rseg to purge queue. */
+        return trx_sys.rseg_array[lhs].last_trx_no() >
+               trx_sys.rseg_array[rhs].last_trx_no();
+      }
+    };
+    byte *begin() { return m_array; }
+    byte *end() { return m_end; }
+    const byte *c_begin() const { return m_array; }
+    const byte *c_end() const { return m_end; }
+    size_t size() const
+    {
+      size_t s= c_end() - c_begin();
+      ut_ad(s <= TRX_SYS_N_RSEGS);
+      return s;
+    }
+    bool empty() const { return !size(); }
+    void clear() { m_end= m_array; }
+
+    /** Push index of trx_sys.rseg_array into min-heap.
+    @param i index to push */
+    void push_rseg_index(byte i)
+    {
+      ut_ad(i < TRX_SYS_N_RSEGS);
+      ut_ad(size() + 1 <= TRX_SYS_N_RSEGS);
+      *m_end++= i;
+      std::push_heap(begin(), end(), trx_rseg_cmp());
+    }
+
+    /** Push rseg to priority queue.
+    @param rseg trx_rseg_t pointer to push */
+    void push(const trx_rseg_t *rseg)
+    {
+      ut_ad(rseg >= trx_sys.rseg_array);
+      ut_ad(rseg < trx_sys.rseg_array + TRX_SYS_N_RSEGS);
+      byte i= byte(rseg - trx_sys.rseg_array);
+      push_rseg_index(i);
+    }
+
+    /** Pop rseg from priority queue.
+    @return pointer to popped trx_rseg_t object */
+    trx_rseg_t *pop()
+    {
+      ut_ad(!empty());
+      byte index= m_array[0];
+      ut_ad(index < TRX_SYS_N_RSEGS);
+      std::pop_heap(begin(), end(), trx_rseg_cmp());
+      --m_end;
+      return &trx_sys.rseg_array[index];
+    }
```

Comment:
If I understood https://en.cppreference.com/w/cpp/algorithm/pop_heap correctly, this could be written as follows, possibly resulting in slightly improved code:
```c++
      ut_ad(!empty());
      std::pop_heap(begin(), end(), trx_seg_cmp());
      byte i= m_array[--m_end];
      ut_ad(i < TRX_SYS_N_RSEGS);
      return &trx_sys.rseg_array[i];